<a href="https://colab.research.google.com/github/nswapnil31/Data-Science-Projects/blob/main/Uber_Partner_Business_Modeling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

import matplotlib.pyplot as plt
import plotly.express as px

**Scenario 1:** It is going to be a huge Saturday and there will need to be many more cars on the road than last week. In order to get drivers to go online, we're assessing the following two bonus options in terms of cost:

* Option 1: $50 for each driver that is online at least 8 hours, accepts 90% of requests, completes 10 trips, and has a rating of 4.7 or better during the time frame;

* Option 2: $4/trip for all drivers who complete 12 trips, and have a 4.7 or better rating.
Using the dataset provided and given Scenario 1, provide answers to the questions below:

1. How much would the total bonus payout be with Option 1?
2. How much would the total bonus payout be with Option 2?
3. How many drivers would qualify for a bonus under Option 1 but not under Option 2?
4. What percentages of drivers online completed less than 10 trips, had an acceptance rate of less than 90%, and had a rating of 4.7 or higher?

In [45]:
df = pd.read_csv('/content/sample_data/dataset_2.csv')
df.head()

Name  Trips Completed Accept Rate  Supply Hours  Rating
0    Abdul                1        100%             3     4.8
1  Abraham               12         83%             5     4.7
2  Adelina                1        100%             2     4.7
3   Akilah                1        100%             2     4.9
4     Alec               21         76%            11     5.0

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119 entries, 0 to 118
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             119 non-null    object 
 1   Trips Completed  119 non-null    int64  
 2   Accept Rate      119 non-null    object 
 3   Supply Hours     119 non-null    int64  
 4   Rating           119 non-null    float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.8+ KB


In [47]:
# convert accept rate to float from string to force conditions later
df['Accept Rate'] = df['Accept Rate'].apply(lambda x: float(x[:-1]))
df

Name  Trips Completed  Accept Rate  Supply Hours  Rating
0       Abdul                1        100.0             3     4.8
1     Abraham               12         83.0             5     4.7
2     Adelina                1        100.0             2     4.7
3      Akilah                1        100.0             2     4.9
4        Alec               21         76.0            11     5.0
..        ...              ...          ...           ...     ...
114    Virgen                4        100.0             6     4.5
115      Yang                7         71.0             2     4.5
116  Yessenia                8         88.0             5     5.0
117    Yukiko                9         78.0             6     4.5
118       Zoe               21         86.0             9     4.9

[119 rows x 5 columns]

In [48]:
df = df.rename(columns={'Trips Completed':'Trips_Completed','Accept Rate': 'Accept_Rate','Supply Hours': 'Supply_Hours'})


In [50]:
df_Option_1=df.query("Supply_Hours>=8 and Accept_Rate>=90 and Trips_Completed>=10 and Rating>=4.7")
df_Option_1

Name  Trips_Completed  Accept_Rate  Supply_Hours  Rating
11       Byron               15        100.0            10     4.9
12     Carlota               14        100.0             8     5.0
19    Dannette               14        100.0             9     4.9
23   Demetrius               14        100.0             9     5.0
26      Dimple               15        100.0            10     4.9
32        Emil               12        100.0             9     5.0
37       Garth               15        100.0            10     5.0
40        Hanh               14         94.0             9     4.9
53      Keshia               20        100.0            11     4.8
57     Latonia               13        100.0             9     5.0
68      Luanne               12         96.0            11     4.9
69     Lyndsey               13         94.0            11     4.9
70     Lynwood               21         95.0            10     5.0
75         Man               23        100.0            10     5.0
76      Marica               16         90.0            10     5.0
84     Natalie               16         94.0             8     4.9
88        Oren               11         91.0             9     4.8
92     Phyllis               10         90.0             8     4.8
95    Salvador               15        100.0            11     4.8
98     Shandra               14        100.0             8     4.9
113     Vashti               12         92.0            11     4.9

In [52]:
df.query("Supply_Hours>=8 and Accept_Rate>=90 and Trips_Completed>=10 and Rating>=4.7").shape[0]*50

1050

How much would the total bonus payout be with Option 1? : 1050






--------------------------------------------------------------------------------


In [53]:
df_Option_2=df.query("Trips_Completed>=12 and Rating>=4.7")

In [54]:
df.query("Trips_Completed>=12 and Rating>=4.7")['Trips_Completed'].sum()*4

2976

How much would the total bonus payout be with Option 2? : **2976**



--------------------------------------------------------------------------------


In [55]:
df_common=df_Option_1[~df_Option_1['Name'].isin(df_Option_2['Name'])]
df_common

Name  Trips_Completed  Accept_Rate  Supply_Hours  Rating
88     Oren               11         91.0             9     4.8
92  Phyllis               10         90.0             8     4.8

How many drivers would qualify for a bonus under Option 1 but not under Option 2: 2



--------------------------------------------------------------------------------


In [58]:
less_trips_df=df.query("Accept_Rate<90 and Trips_Completed<10 and Rating>=4.7")

In [59]:
print(str(len(less_trips_df) / len(df) * 100)+"%")

10.92436974789916%


What percentages of drivers online completed less than 10 trips, had an acceptance rate of less than 90%, and had a rating of 4.7 or higher? : 10.12%



--------------------------------------------------------------------------------


**Scenario 2:** A taxi driver currently generates $200 per day in fares (before expenses), works six days a week, takes three weeks off, and has the following expenses:

* Gas - $200 per week

* Insurance - $400 per month

* Vehicle rent (by the week) - $500

The driver doesn't pay gas and rent expenses on off weeks.

Now, let's assume that the same driver would buy a Town Car and partner with Uber. If he does, his gas expenses would go up by 5%, his insurance expense would decrease by 20%, and he would no longer be renting a vehicle. However, he would need to buy a car. The driver would still take three weeks off per year.

Given Scenario 2, provide answers to the questions below:

5. How much money (after expenses) does the taxi driver make per year without partnering with Uber?
6. You are convincing the same driver above to buy a Town Car and partner with Uber. Assuming the new car is 40.000 USD, how much would the driver's gross fares need to increase per week to fully pay for the car in year 1 and maintain the same yearly profit margin as before?

In [60]:
total_weeks_per_year = 52
weeks_off = 3
fare_per_day = 200 
workday_per_week = 6
total_months_per_year = 12
# expenses
gas_per_week = 200
insurance_per_month = 400
vehicle_rent_by_week = 500

In [61]:
total_expenses = (gas_per_week + vehicle_rent_by_week) * (total_weeks_per_year - weeks_off) + insurance_per_month * total_months_per_year
print("$"+str(total_expenses))

$39100


In [62]:
total_revenue = (total_weeks_per_year - weeks_off) * workday_per_week * fare_per_day
print("$"+str(total_revenue))

$58800


In [63]:
profit_margin = total_revenue - total_expenses
print("$"+str(profit_margin))

$19700


How much money (after expenses) does the taxi driver make per year without partnering with Uber? : $19700

In [64]:
# calculate new expenses
gas_per_week = gas_per_week * 1.05
insurance_per_month = insurance_per_month * 0.8
new_car = 40000

new_total_expenses = new_car + gas_per_week * (total_weeks_per_year - weeks_off) + insurance_per_month * total_months_per_year
print("$"+str(new_total_expenses))

$54130.0


In [65]:
# profit margin is the total revenue from previous question minus new expenses
new_profit_margin = total_revenue - new_total_expenses
print("$"+str(new_profit_margin))

$4670.0


In [66]:
# calculate the weekly fare increase
fare_increase = (profit_margin - new_profit_margin) / (total_weeks_per_year - weeks_off)
print("$"+str(round(fare_increase, 2)))

$306.73


 how much would the driver's gross fares need to increase per week to fully pay for the car in year 1 and maintain the same yearly profit margin as before? : **$306.73**